### PR TITLE
check for embedded NUL char in jl_load (for include)

### PR DESCRIPTION
**Pull request by stevengj**
This fixes a bug I noticed in #16218, in which `include("foo.jl\0")` would erroneously succeed if `"foo.jl"` existed, because the filename would be silently truncated at the NUL char (similar to #10991).

I did an audit of the other functions in `src/*.h`, and I couldn't find any other problems of this kind.

### CODE REVIEW

File: test/loading.jl [hunk 1]
```diff
@@ -4,6 +4,8 @@ using Base.Test
 
 @test @__LINE__ == 5
 
+@test_throws ArgumentError include("test_sourcepath.jl\0")
```
Comment (vtjnash) [inline test/loading.jl]: Maybe just call Core.include directly here? Otherwise, include_from_node1 is confusing the tests.

Comment (stevengj) [inline test/loading.jl]: Okay.

File: src/ast.c [hunk 2]
```diff
@@ -731,6 +731,7 @@ jl_value_t *jl_parse_eval_all(const char *fname, size_t len,
         fl_free_gc_handles(fl_ctx, 1);
     }
     else {
+        assert(memchr(fname, 0, len) == NULL); // was checked already in jl_load
```
Comment (tkelman) [inline src/ast.c]: On release-0.4, `jl-parse-file` is used inside `jl_start_parsing_file` which does not take the `len` as input. Could you propose a backport, preferably w.r.t. tk/backports-0.4.6 if you want to get this in?

Comment (yuyichao) [inline src/ast.c]: FWIW, these functions should not take a length argument since they really don't support anything other than NUL terminated strings....

Comment (tkelman) [inline src/ast.c]: bump again, would need assistance to resolve the conflict if you want this on release-0.4

**Pull request merged**